## Sprint 10. Web scraping i automatització.

### Exercici 1
#### Realitza web scraping de dues de les tres pàgines web proposades utilitzant BeautifulSoup primer i Selenium després. 

- http://quotes.toscrape.com

- https://www.bolsamadrid.es

- www.wikipedia.es (fes alguna cerca primer i escrapeja algun contingut)

In [1]:
import pandas as pd

### Primera web http://quotes.toscrape.com

### Beautiful Soup.
Beautiful Soup es una librería Python que permite extraer información de contenido en formato HTML o XML.

In [2]:
# librerías de Beautiful Soup
import requests
from bs4 import BeautifulSoup

Hemos elegido la web http://quotes.toscrape.com para hacer este proyecto de web scraping, es una web estática, el servidor que almacena la web nos devolverá un documento HTML que contendrá los datos que vemos cuando navegamos por la web.

In [3]:
URL = "http://quotes.toscrape.com"
page = requests.get(URL)

print(page.text)

<!DOCTYPE html>
<html lang="en">
<head>
	<meta charset="UTF-8">
	<title>Quotes to Scrape</title>
    <link rel="stylesheet" href="/static/bootstrap.min.css">
    <link rel="stylesheet" href="/static/main.css">
</head>
<body>
    <div class="container">
        <div class="row header-box">
            <div class="col-md-8">
                <h1>
                    <a href="/" style="text-decoration: none">Quotes to Scrape</a>
                </h1>
            </div>
            <div class="col-md-4">
                <p>
                
                    <a href="/login">Login</a>
                
                </p>
            </div>
        </div>
    

<div class="row">
    <div class="col-md-8">

    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
        <span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
        <span>by <small class="author" itempr

Este es el código de nuestra web con una buena presentación para una fácil lectura.

In [ ]:
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <title>Quotes to Scrape</title>
        <link rel="stylesheet" href="/static/bootstrap.min.css" />
        <link rel="stylesheet" href="/static/main.css" />
    </head>
    <body>
        <div class="container">
            <div class="row header-box">
                <div class="col-md-8">
                    <h1>
                        <a href="/" style="text-decoration: none;">Quotes to Scrape</a>
                    </h1>
                </div>
                <div class="col-md-4">
                    <p>
                        <a href="/login">Login</a>
                    </p>
                </div>
            </div>

            <div class="row">
                <div class="col-md-8">
                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
                        <span>
                            by <small class="author" itemprop="author">Albert Einstein</small>
                            <a href="/author/Albert-Einstein">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="change,deep-thoughts,thinking,world" / >

                            <a class="tag" href="/tag/change/page/1/">change</a>

                            <a class="tag" href="/tag/deep-thoughts/page/1/">deep-thoughts</a>

                            <a class="tag" href="/tag/thinking/page/1/">thinking</a>

                            <a class="tag" href="/tag/world/page/1/">world</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“It is our choices, Harry, that show what we truly are, far more than our abilities.”</span>
                        <span>
                            by <small class="author" itemprop="author">J.K. Rowling</small>
                            <a href="/author/J-K-Rowling">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="abilities,choices" / >

                            <a class="tag" href="/tag/abilities/page/1/">abilities</a>

                            <a class="tag" href="/tag/choices/page/1/">choices</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”</span>
                        <span>
                            by <small class="author" itemprop="author">Albert Einstein</small>
                            <a href="/author/Albert-Einstein">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="inspirational,life,live,miracle,miracles" / >

                            <a class="tag" href="/tag/inspirational/page/1/">inspirational</a>

                            <a class="tag" href="/tag/life/page/1/">life</a>

                            <a class="tag" href="/tag/live/page/1/">live</a>

                            <a class="tag" href="/tag/miracle/page/1/">miracle</a>

                            <a class="tag" href="/tag/miracles/page/1/">miracles</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”</span>
                        <span>
                            by <small class="author" itemprop="author">Jane Austen</small>
                            <a href="/author/Jane-Austen">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="aliteracy,books,classic,humor" / >

                            <a class="tag" href="/tag/aliteracy/page/1/">aliteracy</a>

                            <a class="tag" href="/tag/books/page/1/">books</a>

                            <a class="tag" href="/tag/classic/page/1/">classic</a>

                            <a class="tag" href="/tag/humor/page/1/">humor</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“Imperfection is beauty, madness is genius and it&#39;s better to be absolutely ridiculous than absolutely boring.”</span>
                        <span>
                            by <small class="author" itemprop="author">Marilyn Monroe</small>
                            <a href="/author/Marilyn-Monroe">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="be-yourself,inspirational" / >

                            <a class="tag" href="/tag/be-yourself/page/1/">be-yourself</a>

                            <a class="tag" href="/tag/inspirational/page/1/">inspirational</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“Try not to become a man of success. Rather become a man of value.”</span>
                        <span>
                            by <small class="author" itemprop="author">Albert Einstein</small>
                            <a href="/author/Albert-Einstein">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="adulthood,success,value" / >

                            <a class="tag" href="/tag/adulthood/page/1/">adulthood</a>

                            <a class="tag" href="/tag/success/page/1/">success</a>

                            <a class="tag" href="/tag/value/page/1/">value</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“It is better to be hated for what you are than to be loved for what you are not.”</span>
                        <span>
                            by <small class="author" itemprop="author">André Gide</small>
                            <a href="/author/Andre-Gide">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="life,love" / >

                            <a class="tag" href="/tag/life/page/1/">life</a>

                            <a class="tag" href="/tag/love/page/1/">love</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“I have not failed. I&#39;ve just found 10,000 ways that won&#39;t work.”</span>
                        <span>
                            by <small class="author" itemprop="author">Thomas A. Edison</small>
                            <a href="/author/Thomas-A-Edison">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="edison,failure,inspirational,paraphrased" / >

                            <a class="tag" href="/tag/edison/page/1/">edison</a>

                            <a class="tag" href="/tag/failure/page/1/">failure</a>

                            <a class="tag" href="/tag/inspirational/page/1/">inspirational</a>

                            <a class="tag" href="/tag/paraphrased/page/1/">paraphrased</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“A woman is like a tea bag; you never know how strong it is until it&#39;s in hot water.”</span>
                        <span>
                            by <small class="author" itemprop="author">Eleanor Roosevelt</small>
                            <a href="/author/Eleanor-Roosevelt">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="misattributed-eleanor-roosevelt" / >

                            <a class="tag" href="/tag/misattributed-eleanor-roosevelt/page/1/">misattributed-eleanor-roosevelt</a>
                        </div>
                    </div>

                    <div class="quote" itemscope itemtype="http://schema.org/CreativeWork">
                        <span class="text" itemprop="text">“A day without sunshine is like, you know, night.”</span>
                        <span>
                            by <small class="author" itemprop="author">Steve Martin</small>
                            <a href="/author/Steve-Martin">(about)</a>
                        </span>
                        <div class="tags">
                            Tags: <meta class="keywords" itemprop="keywords" content="humor,obvious,simile" / >

                            <a class="tag" href="/tag/humor/page/1/">humor</a>

                            <a class="tag" href="/tag/obvious/page/1/">obvious</a>

                            <a class="tag" href="/tag/simile/page/1/">simile</a>
                        </div>
                    </div>

                    <nav>
                        <ul class="pager">
                            <li class="next">
                                <a href="/page/2/">Next <span aria-hidden="true">&rarr;</span></a>
                            </li>
                        </ul>
                    </nav>
                </div>
                <div class="col-md-4 tags-box">
                    <h2>Top Ten tags</h2>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 28px;" href="/tag/love/">love</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 26px;" href="/tag/inspirational/">inspirational</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 26px;" href="/tag/life/">life</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 24px;" href="/tag/humor/">humor</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 22px;" href="/tag/books/">books</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 14px;" href="/tag/reading/">reading</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 10px;" href="/tag/friendship/">friendship</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 8px;" href="/tag/friends/">friends</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 8px;" href="/tag/truth/">truth</a>
                    </span>

                    <span class="tag-item">
                        <a class="tag" style="font-size: 6px;" href="/tag/simile/">simile</a>
                    </span>
                </div>
            </div>
        </div>
        <footer class="footer">
            <div class="container">
                <p class="text-muted">Quotes by: <a href="https://www.goodreads.com/quotes">GoodReads.com</a></p>
                <p class="copyright">Made with <span class="sh-red">❤</span> by <a href="https://scrapinghub.com">Scrapinghub</a></p>
            </div>
        </footer>
    </body>
</html>


Para poder trabajar con el código anterior es necesario especificar un parser, que es responsable de transformar un documento HTML o XML en un árbol complejo de objetos Python. Esto permite, por ejemplo, que podamos interactuar con los elementos de una página web.

In [4]:
soup = BeautifulSoup(page.content, "html.parser")

In [7]:
# buscamos el contenido que necesitamos dentro de nuestro código HTML
results = soup.find(class_='container')

In [8]:
type(results)

bs4.element.Tag

Nos centramos en obtener el elemento que contiene la información de cada cita.

In [9]:
# trabajamos sobre la variable que contiene nuestra web y usamos un find_all para obtener un elemento iterable 
# que contenga todas nuestras citas
job_elements = results.find_all('div', class_='quote')

for job_element in job_elements:
    print(job_element, end="\n"*2)  # vemos nuestras citas

<div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
<span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
<span>by <small class="author" itemprop="author">Albert Einstein</small>
<a href="/author/Albert-Einstein">(about)</a>
</span>
<div class="tags">
            Tags:
            <meta class="keywords" content="change,deep-thoughts,thinking,world" itemprop="keywords"/>
<a class="tag" href="/tag/change/page/1/">change</a>
<a class="tag" href="/tag/deep-thoughts/page/1/">deep-thoughts</a>
<a class="tag" href="/tag/thinking/page/1/">thinking</a>
<a class="tag" href="/tag/world/page/1/">world</a>
</div>
</div>

<div class="quote" itemscope="" itemtype="http://schema.org/CreativeWork">
<span class="text" itemprop="text">“It is our choices, Harry, that show what we truly are, far more than our abilities.”</span>
<span>by <small class="author" itemprop="author">J.K. 

En este punto obtenemos la información de cada una de las citas, que son objetos en sí mismos.

In [12]:
# con find() nos movemos dentro del objeto para obtener información detallada de cada parte del elemento
for job_element in job_elements:
    text_element = job_element.find("span", class_="text")
    author_element = job_element.find("small", class_="author")
    link_element = job_element
    keywords_element = job_element.find("meta", class_="keywords")
   
    print(text_element)
    print(author_element)
    print(link_element.find('a')['href'])
    print(keywords_element)
    print()

<span class="text" itemprop="text">“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”</span>
<small class="author" itemprop="author">Albert Einstein</small>
/author/Albert-Einstein
<meta class="keywords" content="change,deep-thoughts,thinking,world" itemprop="keywords"/>

<span class="text" itemprop="text">“It is our choices, Harry, that show what we truly are, far more than our abilities.”</span>
<small class="author" itemprop="author">J.K. Rowling</small>
/author/J-K-Rowling
<meta class="keywords" content="abilities,choices" itemprop="keywords"/>

<span class="text" itemprop="text">“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”</span>
<small class="author" itemprop="author">Albert Einstein</small>
/author/Albert-Einstein
<meta class="keywords" content="inspirational,life,live,miracle,miracles" itemprop="keywords"/>

<span class="text" i

In [11]:
# Vamos a almacenar los datos de las citas en listas y ver cómo se pueden ver los objetos como texto.
text_list = []
author_list = []
link_list = []
keyword = []
keyword_list = []

for job_element in job_elements:
    text_element = job_element.find("span", class_="text")
    text_list.append(text_element.text)
    author_element = job_element.find("small", class_="author")
    author_list.append(author_element.text)
    link_element = job_element
    link_list.append("http://quotes.toscrape.com"+link_element.find('a')['href'])
    keywords_element = job_element.find("meta", class_="keywords")
    keyword.append(keywords_element['content'])
    
    keys = []
    keys = keywords_element['content']
    keyword_list.append(keys)
   
    print(text_element.text)
    print(author_element.text)
    print(link_element.find('a')['href'])
    print(keywords_element['content'])
    print()

“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
Albert Einstein
/author/Albert-Einstein
change,deep-thoughts,thinking,world

“It is our choices, Harry, that show what we truly are, far more than our abilities.”
J.K. Rowling
/author/J-K-Rowling
abilities,choices

“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”
Albert Einstein
/author/Albert-Einstein
inspirational,life,live,miracle,miracles

“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”
Jane Austen
/author/Jane-Austen
aliteracy,books,classic,humor

“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”
Marilyn Monroe
/author/Marilyn-Monroe
be-yourself,inspirational

“Try not to become a man of success. Rather become a man of value.”
Albert Einstein
/author/Albert-Einstein
ad

In [13]:
# pasamos las listas a un dataframe
dict_quotes = {'Cita': text_list,'Autor': author_list,'Link Autor': link_list,'Keywords': keyword}

df_quotes = pd.DataFrame(dict_quotes)
df_quotes

Cita              Autor  \
0  “The world as we have created it is a process ...    Albert Einstein   
1  “It is our choices, Harry, that show what we t...       J.K. Rowling   
2  “There are only two ways to live your life. On...    Albert Einstein   
3  “The person, be it gentleman or lady, who has ...        Jane Austen   
4  “Imperfection is beauty, madness is genius and...     Marilyn Monroe   
5  “Try not to become a man of success. Rather be...    Albert Einstein   
6  “It is better to be hated for what you are tha...         André Gide   
7  “I have not failed. I've just found 10,000 way...   Thomas A. Edison   
8  “A woman is like a tea bag; you never know how...  Eleanor Roosevelt   
9  “A day without sunshine is like, you know, nig...       Steve Martin   

                                          Link Autor  \
0  http://quotes.toscrape.com/author/Albert-Einstein   
1      http://quotes.toscrape.com/author/J-K-Rowling   
2  http://quotes.toscrape.com/author/Albert-Einstein   
3      http://quotes.toscrape.com/author/Jane-Austen   
4   http://quotes.toscrape.com/author/Marilyn-Monroe   
5  http://quotes.toscrape.com/author/Albert-Einstein   
6       http://quotes.toscrape.com/author/Andre-Gide   
7  http://quotes.toscrape.com/author/Thomas-A-Edison   
8  http://quotes.toscrape.com/author/Eleanor-Roos...   
9     http://quotes.toscrape.com/author/Steve-Martin   

                                   Keywords  
0       change,deep-thoughts,thinking,world  
1                         abilities,choices  
2  inspirational,life,live,miracle,miracles  
3             aliteracy,books,classic,humor  
4                 be-yourself,inspirational  
5                   adulthood,success,value  
6                                 life,love  
7  edison,failure,inspirational,paraphrased  
8           misattributed-eleanor-roosevelt  
9                      humor,obvious,simile

In [14]:
# definimos una función que nos permita obtener las citas de cualquier página de nuestra web
def get_pages(types, kind):
    web = f'http://quotes.toscrape.com/{types}/{kind}/'
    page = requests.get(web)
    soup = BeautifulSoup(page.content, "html.parser")
    results = soup.find(id='')
    job_elements = results.find_all('div', class_='quote')
    
    text_list = []
    author_list = []
    link_list = []
    keyword = []
    
    for job_element in job_elements:
        text_element = job_element.find("span", class_="text")
        text_list.append(text_element.text)
        author_element = job_element.find("small", class_="author")
        author_list.append(author_element.text)
        link_element = job_element
        link_list.append("http://quotes.toscrape.com"+link_element.find('a')['href'])
        keywords_element = job_element.find("meta", class_="keywords")
        keyword.append(keywords_element['content'])

    dict_quotes = {'Cita': text_list,'Autor': author_list,'Link Autor': link_list,'Keywords': keyword}

    df = pd.DataFrame(dict_quotes)
    
    return df

In [15]:
# definimos un dataframe que vaya contatenando los dataframes generados en cada página en la que entremos
df_quotes_complete = pd.DataFrame(columns=('Cita','Autor','Link Autor','Keywords'))

# recorremos las páginas de nuestra web
kinds = list(range(1,11))
page = 'page'
for kind in kinds:
    df_quotes_complete = pd.concat((df_quotes_complete, get_pages(page,kind)), ignore_index=True)

In [ ]:
# otra manera de hacerlo
#df_quotes_page = [get_pages(page,kind) for kind in kinds] # hacemos una lista de dataframes
#df_quotes_complete = pd.concat(df_quotes_page, ignore_index=True)  # la función concat() admite una lista

In [16]:
df_quotes_complete

Cita               Autor  \
0   “The world as we have created it is a process ...     Albert Einstein   
1   “It is our choices, Harry, that show what we t...        J.K. Rowling   
2   “There are only two ways to live your life. On...     Albert Einstein   
3   “The person, be it gentleman or lady, who has ...         Jane Austen   
4   “Imperfection is beauty, madness is genius and...      Marilyn Monroe   
..                                                ...                 ...   
95  “You never really understand a person until yo...          Harper Lee   
96  “You have to write the book that wants to be w...   Madeleine L'Engle   
97  “Never tell the truth to people who are not wo...          Mark Twain   
98        “A person's a person, no matter how small.”           Dr. Seuss   
99  “... a mind needs books as a sword needs a whe...  George R.R. Martin   

                                           Link Autor  \
0   http://quotes.toscrape.com/author/Albert-Einstein   
1       http://quotes.toscrape.com/author/J-K-Rowling   
2   http://quotes.toscrape.com/author/Albert-Einstein   
3       http://quotes.toscrape.com/author/Jane-Austen   
4    http://quotes.toscrape.com/author/Marilyn-Monroe   
..                                                ...   
95       http://quotes.toscrape.com/author/Harper-Lee   
96  http://quotes.toscrape.com/author/Madeleine-LE...   
97       http://quotes.toscrape.com/author/Mark-Twain   
98         http://quotes.toscrape.com/author/Dr-Seuss   
99  http://quotes.toscrape.com/author/George-R-R-M...   

                                             Keywords  
0                 change,deep-thoughts,thinking,world  
1                                   abilities,choices  
2            inspirational,life,live,miracle,miracles  
3                       aliteracy,books,classic,humor  
4                           be-yourself,inspirational  
..                                                ...  
95                                better-life-empathy  
96  books,children,difficult,grown-ups,write,write...  
97                                              truth  
98                                      inspirational  
99                                         books,mind  

[100 rows x 4 columns]

In [17]:
df_quotes_complete['Cita'].nunique()

100

In [18]:
df_quotes_complete['Autor'].nunique()

50

In [19]:
df_quotes_complete.to_csv('quotes_scraping_BS4.csv', index = False)  # guardamos el archivo

In [20]:
keyword_list = []  
keywords = df_quotes_complete['Keywords'] # hacemos una lista con las keywords
for key in keywords:  # las separamos
    key = key.split(',')
    keyword_list+=key
print(keyword_list)
set(keyword_list)  # eliminamos las repetidas

['change', 'deep-thoughts', 'thinking', 'world', 'abilities', 'choices', 'inspirational', 'life', 'live', 'miracle', 'miracles', 'aliteracy', 'books', 'classic', 'humor', 'be-yourself', 'inspirational', 'adulthood', 'success', 'value', 'life', 'love', 'edison', 'failure', 'inspirational', 'paraphrased', 'misattributed-eleanor-roosevelt', 'humor', 'obvious', 'simile', 'friends', 'heartbreak', 'inspirational', 'life', 'love', 'sisters', 'courage', 'friends', 'simplicity', 'understand', 'love', 'fantasy', 'life', 'navigation', 'activism', 'apathy', 'hate', 'indifference', 'inspirational', 'love', 'opposite', 'philosophy', 'friendship', 'lack-of-friendship', 'lack-of-love', 'love', 'marriage', 'unhappy-marriage', 'books', 'contentment', 'friends', 'friendship', 'life', 'fate', 'life', 'misattributed-john-lennon', 'planning', 'plans', 'love', 'poetry', 'happiness', 'attributed-no-source', 'humor', 'religion', 'humor', 'comedy', 'life', 'yourself', 'children', 'fairy-tales', '', 'imagination

{'',
 'abilities',
 'activism',
 'adulthood',
 'adventure',
 'age',
 'alcohol',
 'aliteracy',
 'apathy',
 'attributed',
 'attributed-no-source',
 'authors',
 'be-yourself',
 'beatles',
 'better-life-empathy',
 'bilbo',
 'books',
 'change',
 'children',
 'chocolate',
 'choices',
 'christianity',
 'classic',
 'comedy',
 'connection',
 'contentment',
 'courage',
 'death',
 'deep-thoughts',
 'difficult',
 'dreamers',
 'dreaming',
 'dreams',
 'drug',
 'dumbledore',
 'edison',
 'education',
 'elizabeth-bennet',
 'failure',
 'fairy-tales',
 'fairytales',
 'faith',
 'fantasy',
 'fate',
 'fear',
 'food',
 'friends',
 'friendship',
 'girls',
 'god',
 'good',
 'growing-up',
 'grown-ups',
 'happiness',
 'hate',
 'heartbreak',
 'hope',
 'humor',
 'imagination',
 'indifference',
 'insanity',
 'inspiration',
 'inspirational',
 'integrity',
 'jane-austen',
 'journey',
 'knowledge',
 'lack-of-friendship',
 'lack-of-love',
 'learning',
 'library',
 'lies',
 'life',
 'literature',
 'live',
 'live-death-l

In [21]:
keyword_finales = list(set(keyword_list))
print(keyword_finales)

['', 'seuss', 'fairy-tales', 'planning', 'classic', 'navigation', 'dreaming', 'wisdom', 'faith', 'adventure', 'quest', 'comedy', 'miracle', 'food', 'open-mind', 'age', 'integrity', 'writers', 'understand', 'philosophy', 'plans', 'regrets', 'humor', 'better-life-empathy', 'difficult', 'thought', 'reading-books', 'live', 'christianity', 'unhappy-marriage', 'elizabeth-bennet', 'heartbreak', 'misattributed-john-lennon', 'lack-of-love', 'marriage', 'bilbo', 'lost', 'good', 'reading', 'mind', 'courage', 'knowledge', 'miracles', 'misattributed-to-mother-teresa', 'choices', 'read', 'aliteracy', 'yourself', 'live-death-love', 'self-indulgence', 'peace', 'fear', 'mistakes', 'fairytales', 'writing', 'fate', 'dumbledore', 'obvious', 'girls', 'women', 'books', 'deep-thoughts', 'alcohol', 'abilities', 'wander', 'success', 'the-hunger-games', 'god', 'life', 'music', 'sisters', 'world', 'poetry', 'inspiration', 'opposite', 'tea', 'attributed-no-source', 'authors', 'lies', 'journey', 'misattributed-ele

In [22]:
keyword_finales = keyword_finales[1:]   # ahora tenemos la lista de las keywords de nuestra web

keyword_finales

['seuss',
 'fairy-tales',
 'planning',
 'classic',
 'navigation',
 'dreaming',
 'wisdom',
 'faith',
 'adventure',
 'quest',
 'comedy',
 'miracle',
 'food',
 'open-mind',
 'age',
 'integrity',
 'writers',
 'understand',
 'philosophy',
 'plans',
 'regrets',
 'humor',
 'better-life-empathy',
 'difficult',
 'thought',
 'reading-books',
 'live',
 'christianity',
 'unhappy-marriage',
 'elizabeth-bennet',
 'heartbreak',
 'misattributed-john-lennon',
 'lack-of-love',
 'marriage',
 'bilbo',
 'lost',
 'good',
 'reading',
 'mind',
 'courage',
 'knowledge',
 'miracles',
 'misattributed-to-mother-teresa',
 'choices',
 'read',
 'aliteracy',
 'yourself',
 'live-death-love',
 'self-indulgence',
 'peace',
 'fear',
 'mistakes',
 'fairytales',
 'writing',
 'fate',
 'dumbledore',
 'obvious',
 'girls',
 'women',
 'books',
 'deep-thoughts',
 'alcohol',
 'abilities',
 'wander',
 'success',
 'the-hunger-games',
 'god',
 'life',
 'music',
 'sisters',
 'world',
 'poetry',
 'inspiration',
 'opposite',
 'tea',
 '

In [23]:
len(keyword_finales)

137

En nuestra web hay 137 keywords.

Ahora queremos obtener las citas de la página de la etiqueta "life".

In [24]:
tag = 'tag'
life = 'life'
get_pages(tag, life) # la función nos devuelve un dataframe con las citas de la página "life"

Cita                Autor  \
0  “There are only two ways to live your life. On...      Albert Einstein   
1  “It is better to be hated for what you are tha...           André Gide   
2  “This life is what you make it. No matter what...       Marilyn Monroe   
3  “I may not have gone where I intended to go, b...        Douglas Adams   
4  “Good friends, good books, and a sleepy consci...           Mark Twain   
5  “Life is what happens to us while we are makin...       Allen Saunders   
6  “Today you are You, that is truer than true. T...            Dr. Seuss   
7  “Life is like riding a bicycle. To keep your b...      Albert Einstein   
8  “Life isn't about finding yourself. Life is ab...  George Bernard Shaw   
9  “Finish each day and be done with it. You have...  Ralph Waldo Emerson   

                                          Link Autor  \
0  http://quotes.toscrape.com/author/Albert-Einstein   
1       http://quotes.toscrape.com/author/Andre-Gide   
2   http://quotes.toscrape.com/author/Marilyn-Monroe   
3    http://quotes.toscrape.com/author/Douglas-Adams   
4       http://quotes.toscrape.com/author/Mark-Twain   
5   http://quotes.toscrape.com/author/Allen-Saunders   
6         http://quotes.toscrape.com/author/Dr-Seuss   
7  http://quotes.toscrape.com/author/Albert-Einstein   
8  http://quotes.toscrape.com/author/George-Berna...   
9  http://quotes.toscrape.com/author/Ralph-Waldo-...   

                                            Keywords  
0           inspirational,life,live,miracle,miracles  
1                                          life,love  
2  friends,heartbreak,inspirational,life,love,sis...  
3                                    life,navigation  
4          books,contentment,friends,friendship,life  
5  fate,life,misattributed-john-lennon,planning,p...  
6                               comedy,life,yourself  
7                                        life,simile  
8                        inspirational,life,yourself  
9                                       life,regrets

Haremos un dataframe con las citas de las diez etiquetas que más se repiten.

In [26]:
Top_Ten_tag = ['love', 'inspirational', 'life', 'humor', 'books', 'reading', 'friendship', 'friends', 'truth', 'simile']
df_quotes_complete_tags = pd.DataFrame(columns=('Cita','Autor','Link Autor','Keywords'))
page = 'tag'

for kind in Top_Ten_tag:
    df_quotes_complete_tags = pd.concat((df_quotes_complete_tags, get_pages(page,kind)), ignore_index=True)

In [27]:
df_quotes_complete_tags

Cita                Autor  \
0   “It is better to be hated for what you are tha...           André Gide   
1   “This life is what you make it. No matter what...       Marilyn Monroe   
2   “You may not be her first, her last, or her on...           Bob Marley   
3   “The opposite of love is not hate, it's indiff...          Elie Wiesel   
4   “It is not a lack of love, but a lack of frien...  Friedrich Nietzsche   
..                                                ...                  ...   
68  “The truth." Dumbledore sighed. "It is a beaut...         J.K. Rowling   
69  “Never tell the truth to people who are not wo...           Mark Twain   
70  “A day without sunshine is like, you know, nig...         Steve Martin   
71  “Life is like riding a bicycle. To keep your b...      Albert Einstein   
72                 “He's like a drug for you, Bella.”      Stephenie Meyer   

                                           Link Autor  \
0        http://quotes.toscrape.com/author/Andre-Gide   
1    http://quotes.toscrape.com/author/Marilyn-Monroe   
2        http://quotes.toscrape.com/author/Bob-Marley   
3       http://quotes.toscrape.com/author/Elie-Wiesel   
4   http://quotes.toscrape.com/author/Friedrich-Ni...   
..                                                ...   
68      http://quotes.toscrape.com/author/J-K-Rowling   
69       http://quotes.toscrape.com/author/Mark-Twain   
70     http://quotes.toscrape.com/author/Steve-Martin   
71  http://quotes.toscrape.com/author/Albert-Einstein   
72  http://quotes.toscrape.com/author/Stephenie-Meyer   

                                             Keywords  
0                                           life,love  
1   friends,heartbreak,inspirational,life,love,sis...  
2                                                love  
3   activism,apathy,hate,indifference,inspirationa...  
4   friendship,lack-of-friendship,lack-of-love,lov...  
..                                                ...  
68                                              truth  
69                                              truth  
70                               humor,obvious,simile  
71                                        life,simile  
72                                drug,romance,simile  

[73 rows x 4 columns]

In [29]:
df_quotes_complete_tags['Cita'].nunique()

52

In [30]:
df_quotes_complete_tags['Autor'].nunique()

38

En las páginas de nuestra web que corresponden a las diez palabras clave más repetidas tenemos 52 citas de 38 autores diferentes.

In [33]:
# guardamos el archivo
df_quotes_complete_tags.to_csv('quotes_tags_scraping_BS4.csv', index = False)

### Selenium.

In [54]:
from selenium.webdriver import Firefox
from selenium.webdriver.firefox.service import Service
from selenium.webdriver.support.select import Select
from selenium.webdriver.firefox.options import Options
from selenium.webdriver.common.by import By
import time

In [55]:
opts = Options()
opts.add_argument("--headless")
browser = Firefox(options=opts)  # abrimos el navegador
browser.get(url = 'https://quotes.toscrape.com/')  # contactamos con la web
results = browser.find_element(By.CLASS_NAME, 'container')  # leemos el contenido

In [56]:
print(results.text)

Quotes to Scrape
Login
“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
by Albert Einstein (about)
Tags: change deep-thoughts thinking world
“It is our choices, Harry, that show what we truly are, far more than our abilities.”
by J.K. Rowling (about)
Tags: abilities choices
“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”
by Albert Einstein (about)
Tags: inspirational life live miracle miracles
“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”
by Jane Austen (about)
Tags: aliteracy books classic humor
“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”
by Marilyn Monroe (about)
Tags: be-yourself inspirational
“Try not to become a man of success. Rather become a man of value.”
by Albert Einstein (about)
Tags: adulthood success

In [57]:
# localizamos el elemento con contiene cada cita
message = browser.find_element(by=By.XPATH, value="/html/body/div/div[2]/div[1]/div[1]")  
value = message.text
print(value)

“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
by Albert Einstein (about)
Tags: change deep-thoughts thinking world


In [58]:
# pasamos cada elemento de la cita a una variable
cita = message.find_element(by=By.CLASS_NAME, value="text")
autor = message.find_element(by=By.CLASS_NAME, value="author")
tags = message.find_element(by=By.CLASS_NAME, value="tags")
tags_texto = str(tags.text)[6:]
type(tags_texto)
print(cita.text)
print(autor.text)
print(tags_texto)

“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”
Albert Einstein
change deep-thoughts thinking world


In [59]:
# obtenemos todas las citas de una página y las ordenamos en listas
i = 1
cita_list = []
autor_list = []
tags_list = []
while i<11:
    message = browser.find_element(by=By.XPATH, value=("/html/body/div/div[2]/div[1]/div[{}]").format(i))
    cita_list.append((message.find_element(by=By.CLASS_NAME, value="text").text))
    autor_list.append((message.find_element(by=By.CLASS_NAME, value="author").text))
    tags = message.find_element(by=By.CLASS_NAME, value="tags")
    tags_list.append(str(tags.text)[6:])
    i+=1
print(cita_list)
print(autor_list)
print(tags_list)

['“The world as we have created it is a process of our thinking. It cannot be changed without changing our thinking.”', '“It is our choices, Harry, that show what we truly are, far more than our abilities.”', '“There are only two ways to live your life. One is as though nothing is a miracle. The other is as though everything is a miracle.”', '“The person, be it gentleman or lady, who has not pleasure in a good novel, must be intolerably stupid.”', "“Imperfection is beauty, madness is genius and it's better to be absolutely ridiculous than absolutely boring.”", '“Try not to become a man of success. Rather become a man of value.”', '“It is better to be hated for what you are than to be loved for what you are not.”', "“I have not failed. I've just found 10,000 ways that won't work.”", "“A woman is like a tea bag; you never know how strong it is until it's in hot water.”", '“A day without sunshine is like, you know, night.”']
['Albert Einstein', 'J.K. Rowling', 'Albert Einstein', 'Jane Aus

In [60]:
# hacemos un dataframe con las listas
dict_quotes_Sel = {'Cita': cita_list,'Autor': autor_list,'Keywords': tags_list}

df_quotes_Sel = pd.DataFrame(dict_quotes_Sel)
df_quotes_Sel

Cita              Autor  \
0  “The world as we have created it is a process ...    Albert Einstein   
1  “It is our choices, Harry, that show what we t...       J.K. Rowling   
2  “There are only two ways to live your life. On...    Albert Einstein   
3  “The person, be it gentleman or lady, who has ...        Jane Austen   
4  “Imperfection is beauty, madness is genius and...     Marilyn Monroe   
5  “Try not to become a man of success. Rather be...    Albert Einstein   
6  “It is better to be hated for what you are tha...         André Gide   
7  “I have not failed. I've just found 10,000 way...   Thomas A. Edison   
8  “A woman is like a tea bag; you never know how...  Eleanor Roosevelt   
9  “A day without sunshine is like, you know, nig...       Steve Martin   

                                   Keywords  
0       change deep-thoughts thinking world  
1                         abilities choices  
2  inspirational life live miracle miracles  
3             aliteracy books classic humor  
4                 be-yourself inspirational  
5                   adulthood success value  
6                                 life love  
7  edison failure inspirational paraphrased  
8           misattributed-eleanor-roosevelt  
9                      humor obvious simile

In [61]:
# navegamos por la toda web para obtener todas las citas
cita_list = []
autor_list = []
tags_list = []

cortar = 1
while cortar < 11:
    i=1
    while i<11:
        message = browser.find_element(by=By.XPATH, value=("/html/body/div/div[2]/div[1]/div[{}]").format(i))
        cita_list.append((message.find_element(by=By.CLASS_NAME, value="text").text))
        autor_list.append((message.find_element(by=By.CLASS_NAME, value="author").text))
        tags = message.find_element(by=By.CLASS_NAME, value="tags")
        tags_list.append(str(tags.text)[6:])
        i+=1
    
    # boton para cargar más items
    time.sleep(3)
    if cortar == 1:
        browser.find_element(by=By.XPATH, value='/html/body/div/div[2]/div[1]/nav/ul/li/a').click()
        
    elif cortar <10:
        browser.find_element(by=By.XPATH, value='/html/body/div/div[2]/div[1]/nav/ul/li[2]/a').click() 
                                                
    cortar+=1 

In [62]:
# hacemos un dataframe con toda la información de la web
dict_quotes_Sel_web = {'Cita': cita_list,'Autor': autor_list, 'Keywords': tags_list}

dict_quotes_Sel_web = pd.DataFrame(dict_quotes_Sel_web)

In [64]:
dict_quotes_Sel_web

Cita               Autor  \
0   “The world as we have created it is a process ...     Albert Einstein   
1   “It is our choices, Harry, that show what we t...        J.K. Rowling   
2   “There are only two ways to live your life. On...     Albert Einstein   
3   “The person, be it gentleman or lady, who has ...         Jane Austen   
4   “Imperfection is beauty, madness is genius and...      Marilyn Monroe   
..                                                ...                 ...   
95  “You never really understand a person until yo...          Harper Lee   
96  “You have to write the book that wants to be w...   Madeleine L'Engle   
97  “Never tell the truth to people who are not wo...          Mark Twain   
98        “A person's a person, no matter how small.”           Dr. Seuss   
99  “... a mind needs books as a sword needs a whe...  George R.R. Martin   

                                             Keywords  
0                 change deep-thoughts thinking world  
1                                   abilities choices  
2            inspirational life live miracle miracles  
3                       aliteracy books classic humor  
4                           be-yourself inspirational  
..                                                ...  
95                                better-life-empathy  
96  books children difficult grown-ups write write...  
97                                              truth  
98                                      inspirational  
99                                         books mind  

[100 rows x 3 columns]

In [65]:
dict_quotes_Sel_web.shape

(100, 3)

In [66]:
dict_quotes_Sel_web['Cita'].nunique()

100

In [67]:
dict_quotes_Sel_web.dtypes

Cita        object
Autor       object
Keywords    object
dtype: object

In [68]:
# guardamos los datos en un archivo
dict_quotes_Sel_web.to_csv('quotes_scraping_Sel.csv', index = False)

In [69]:
# cerramos el navegador
browser.close()

### Segunda Web www.wikipedia.es

### Beautiful Soup.

In [77]:
URL = "https://es.wikipedia.org/wiki/Beautiful_Soup"
page = requests.get(URL)

#### Contenido general de la página de Wikipedia correspondiente a la búsqueda Beautifu Soup.

In [78]:
print(page.text)

<!DOCTYPE html>
<html class="client-nojs" lang="es" dir="ltr">
<head>
<meta charset="UTF-8"/>
<title>Beautiful Soup - Wikipedia, la enciclopedia libre</title>
<script>document.documentElement.className="client-js";(function(){var cookie=document.cookie.match(/(?:^|; )eswikimwclientprefs=([^;]+)/);if(cookie){var featureName=cookie[1];document.documentElement.className=document.documentElement.className.replace(featureName+'-enabled',featureName+'-disabled');}}());RLCONF={"wgBreakFrames":false,"wgSeparatorTransformTable":[",\t."," \t,"],"wgDigitTransformTable":["",""],"wgDefaultDateFormat":"dmy","wgMonthNames":["","enero","febrero","marzo","abril","mayo","junio","julio","agosto","septiembre","octubre","noviembre","diciembre"],"wgRequestId":"46dad0af-0a59-4ec3-bab9-0de38ccd7d67","wgCSPNonce":false,"wgCanonicalNamespace":"","wgCanonicalSpecialPageName":false,"wgNamespaceNumber":0,"wgPageName":"Beautiful_Soup","wgTitle":"Beautiful Soup","wgCurRevisionId":146215257,"wgRevisionId":146215257,"

In [79]:
soup = BeautifulSoup(page.content, "html.parser")
results = soup.find(id='content')

#### Título de la búsqueda

In [80]:
job_elements = results.find('h1', class_='firstHeading')
print(job_elements.text)

Beautiful Soup


#### Primera definición de la búsqueda.

In [81]:
job_elements = results.find('div', class_='mw-parser-output').find('p')
print(job_elements.text)

Beautiful Soup es una biblioteca de Python para analizar documentos HTML (incluyendo los que tienen un marcado incorrecto). Esta biblioteca crea un árbol con todos los elementos del documento y puede ser utilizado para extraer información. Por lo tanto, esta biblioteca es útil para realizar web scraping — extraer información de sitios web.[2]​



### Selenium.

In [90]:
opts = Options()
opts.add_argument("--headless")
browser = Firefox(options=opts)
browser.get(url = 'https://es.wikipedia.org/wiki/Selenium')

In [84]:
results = browser.find_element(By.CLASS_NAME, value = 'mw-parser-output')

#### Contenido general de la página de Wikipedia correspondiente a la búsqueda Beautifu Soup.

In [85]:
print(results.text)

Selenium
Parte de Software Freedom Conservancy
Información general
Tipo de programa Automatización de pruebas
Desarrollador SeleniumHQ
Licencia Apache License
Información técnica
Programado en Java
Versiones
Última versión estable 4.1.2 ( 30 de enero de 2022 (1 año y 11 días))
Enlaces
Sitio web oficial
Repositorio de código
[editar datos en Wikidata]
Selenium es un entorno de pruebas de software para aplicaciones basadas en la web. Selenium provee una herramienta de grabar/reproducir para crear pruebas sin usar un lenguaje de scripting para pruebas (Selenium IDE). Incluye también un lenguaje específico de dominio para pruebas (Selenese) para escribir pruebas en un amplio número de lenguajes de programación populares incluyendo Java, C#, Ruby, Groovy, Perl, PHP y Python. Las pruebas pueden ejecutarse entonces usando la mayoría de los navegadores web modernos en diferentes sistemas operativos como Windows, Linux y OSX.
Índice
1 Historia
2 Componentes de Selenium
2.1 Selenium IDE
2.2 Sele

#### Definición obtenida en la página.

In [86]:
contenido = results.find_element(By.TAG_NAME, value = 'p')
print(contenido.text)

Selenium es un entorno de pruebas de software para aplicaciones basadas en la web. Selenium provee una herramienta de grabar/reproducir para crear pruebas sin usar un lenguaje de scripting para pruebas (Selenium IDE). Incluye también un lenguaje específico de dominio para pruebas (Selenese) para escribir pruebas en un amplio número de lenguajes de programación populares incluyendo Java, C#, Ruby, Groovy, Perl, PHP y Python. Las pruebas pueden ejecutarse entonces usando la mayoría de los navegadores web modernos en diferentes sistemas operativos como Windows, Linux y OSX.


In [91]:
browser.close()

### Exercici 2
#### Documenta en un Word el teu conjunt de dades generat amb la informació que tenen els diferents arxius de Kaggle.

### Sobre los dataset obtenidos con Beautiful Soup y https://quotes.toscrape.com/
#### Primer dataset.
Este dataset está disponible en el archivo quotes_scraping_BS4.csv, tiene un tamaño de 21KB.  
Este dataset recoge las citas contenidas en la web https://quotes.toscrape.com/, tiene 100 registros y cuatro columnas tipo objeto:
- Cita: citas extraidas de la web, 100 valores únicos
- Autor: autores de las citas, 50 valores únicos
- Link: link a la información de los autores, 100 valores únicos
- Keywords: palabras representativas de la cita, variable categórica  

#### Segundo dataset.  
Este dataset está disponible en el archivo quotes_tags_scraping_BS4.csv, tiene un tamaño de 19KB.  
Este dataset recoge las citas contenidas en las diez páginas de la web https://quotes.toscrape.com/ que corresponden a las 10 palabras más usadas en la descripción de las citas: love,inspirational,life,humor,books,reading,friendship,friends,truth,simile. Tiene 73 registros y cuatro columnas tipo objeto:
- Cita: citas extraidas de la web, 52 valores únicos
- Autor: autores de las citas, 38 valores únicos
- Link: link a la información de los autores, 100 valores únicos
- Keywords: palabras representativas de la cita, variable categórica  

### Sobre el dataset obtenido con Selenium y https://quotes.toscrape.com/
Este dataset está disponible en el archivo "quotes_scraping_Sel.csv", tiene un tamaño de KB.  
Este dataset recogen las citas contenidas en la web https://quotes.toscrape.com/, tiene 100 registros y cuatro columnas tipo objeto, una de ella categórica nominal:  
- Cita: citas extraidas de la web, 100 valores únicos
- Autor: autores de las citas, 50 valores únicos
- Keywords: palabras representativas de la cita, variable categórica  
